### PR TITLE
Support primitives (ints, strings, JsonLiterals, etc) on a top-level 

### LINKED ISSUE #451: Primitive values cannot be serialized to JsonElement
**Describe the bug**
While `Json.plain.stringify` can be used on primitives and complex objects alike, `Json.plain.toJson` fails on primitive objects with an exception:

```
kotlinx.serialization.SerializationException: No tag in stack for requested element
	at kotlinx.serialization.TaggedEncoder.popTag(Tagged.kt:160)
	at kotlinx.serialization.TaggedEncoder.encodeBoolean(Tagged.kt:95)
	at kotlinx.serialization.internal.BooleanSerializer.serialize(Primitives.kt:54)
	at kotlinx.serialization.internal.BooleanSerializer.serialize(Primitives.kt:51)
	at kotlinx.serialization.json.internal.AbstractJsonTreeOutput.encodeSerializableValue(TreeJsonOutput.kt:174)
	at kotlinx.serialization.CoreKt.encode(Core.kt:73)
	at kotlinx.serialization.json.internal.TreeJsonOutputKt.writeJson(TreeJsonOutput.kt:29)
	at kotlinx.serialization.json.Json.toJson(Json.kt:129)
```

**To Reproduce**
Run (e.g. `./gradlew run`) the attached demo project from
[primitive-serialization.zip](https://github.com/Kotlin/kotlinx.serialization/files/3151857/primitive-serialization.zip)

**Expected behavior**
I would expect both Strings and Booleans to be serialized to JsonPrimitive instances.

**Environment**
 - Kotlin version: 1.3.31
 - Library version: 0.11.0
 - Kotlin platforms: JVM
 - Gradle version: 4.10.3

### PR DESCRIPTION
**Pull request by sandwwraith**
…when saving/restoring JSON AST

Fixes #451

### CODE REVIEW

File: runtime/commonMain/src/kotlinx/serialization/internal/Enums.kt [hunk 1]
```diff
@@ -66,6 +54,14 @@ open class CommonEnumSerializer<T>(val serialName: String, val choices: Array<T>
             { "$index is not among valid $serialName choices, choices size is ${choices.size}" }
         return choices[index]
     }
+
+    companion object {
+        /**
+         * Factory method for easier instantiation from user/test code.
+         */
+        inline fun <reified E: Enum<E>> create(serialName: String): CommonEnumSerializer<E> =
+            CommonEnumSerializer(serialName, enumValues<E>(), enumValues<E>().map { it.name }.toTypedArray())
```
Comment (qwwdfsad) [inline runtime/commonMain/src/kotlinx/serialization/internal/Enums.kt]: 1) Unnecessary type argument in `enumValue`
2) This API is only used in tests, thus probably shouldn't be public
3) More Kotlin-ish API would be constructor-like functions:  `inline fun <reified E: Enum<E>>CommonEnumSerializer(serialName: String)` without companion

File: runtime/commonMain/src/kotlinx/serialization/json/internal/TreeJsonInput.kt [hunk 2]
```diff
@@ -1,27 +1,28 @@
...
-        else -> JsonTreeInput(this, cast(element))
+    val input = when (element) {
+        is JsonObject -> JsonTreeInput(this, element)
+        is JsonArray -> JsonTreeListInput(this, element)
+        is JsonLiteral -> JsonPrimitiveInput(this, element)
+        is JsonNull -> {
+            deserializer.descriptor
+                .also { desc -> require(desc.isNullable) { "Read JsonNull and expected nullable descriptor, but has $desc" } }
+            @Suppress("NULL_FOR_NONNULL_TYPE")
+            return null
```
Comment (qwwdfsad) [inline runtime/commonMain/src/kotlinx/serialization/json/internal/TreeJsonInput.kt]: Semantically, it is an unchecked cast: `@Suppress("UNCHECKED_CAST") return null as T`
I should've written this in the first place

File: runtime/commonTest/src/kotlinx/serialization/json/serializers/JsonNativePrimitivesTest.kt [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+ */
+
+package kotlinx.serialization.json.serializers
+
+import kotlinx.serialization.SampleEnum
+import kotlinx.serialization.internal.*
+import kotlinx.serialization.json.JsonTestBase
+import kotlin.test.Test
+
+class JsonNativePrimitivesTest : JsonTestBase() {
```
Comment (qwwdfsad) [inline runtime/commonTest/src/kotlinx/serialization/json/serializers/JsonNativePrimitivesTest.kt]: Please add tests for nullable primitives with `null` as data as well

File: runtime/commonMain/src/kotlinx/serialization/json/internal/TreeJsonOutput.kt [hunk 4]
```diff
@@ -118,6 +111,28 @@ private sealed class AbstractJsonTreeOutput(
...
+    companion object {
+        const val primitive = "primitive"
+    }
+
+    init {
+        pushTag(primitive)
+    }
+
+    override fun putElement(key: String, element: JsonElement) {
+        require(key == primitive)
```
Comment (qwwdfsad) [inline runtime/commonMain/src/kotlinx/serialization/json/internal/TreeJsonOutput.kt]: 1) [nitpick] you can use referential comparison here :) 
2) Please add a corresponding error message 
3) It is better to use top-level const to avoid blowing the binary with unnecessary empty classes

File: runtime/commonTest/src/kotlinx/serialization/json/serializers/JsonPrimitiveSerializerTest.kt [hunk 5]
```diff
@@ -78,26 +78,26 @@ class JsonPrimitiveSerializerTest : JsonTestBase() {
         assertEquals(JsonPrimitiveWrapper(JsonLiteral("239")), strict.parse(JsonPrimitiveWrapper.serializer(), string, useStreaming))
     }
 
-    @Test // TODO Top-level nulls are not supported in tagged encoders
-    fun testTopLevelPrimitive() {//} = parametrizedTest { useStreaming ->
-        val string = strict.stringify(JsonPrimitiveSerializer, JsonLiteral(42))
+    @Test
```
Comment (qwwdfsad) [inline runtime/commonTest/src/kotlinx/serialization/json/serializers/JsonPrimitiveSerializerTest.kt]: `JsonNullSerializerTest` contains commented code waiting for this change as well

File: runtime/commonMain/src/kotlinx/serialization/json/internal/TreeJsonOutput.kt [hunk 6]
```diff
@@ -68,8 +56,13 @@ private sealed class AbstractJsonTreeOutput(
     }
 
     override fun <T> encodeSerializableValue(serializer: SerializationStrategy<T>, value: T) {
-        encodePolymorphically(serializer, value) {
-            writePolymorphic = true
+        if (serializer.descriptor.kind !is PrimitiveKind && serializer.descriptor.kind !== UnionKind.ENUM_KIND || currentTagOrNull != null) {
+            encodePolymorphically(serializer, value) { writePolymorphic = true }
+        } else {
+            JsonTreePrimitiveOutput(json, nodeConsumer).apply {
```
Comment (qwwdfsad) [inline runtime/commonMain/src/kotlinx/serialization/json/internal/TreeJsonOutput.kt]: IMO, `apply` is not applicable here semantically, because it creates a false sense that you are somehow changing it in `block`, but this is not true. And every indentation level makes code less readable.
E.g.
```
if (serializer.descriptor.kind !is PrimitiveKind && serializer.descriptor.kind !== UnionKind.ENUM_KIND || currentTagOrNull != null) {
    return encodePolymorphically(serializer, value) { writePolymorphic = true }
}
val output = JsonTreePrimitiveOutput(json, nodeConsumer)
output.encodeSerializableValue(serializer, value)
output.endEncode(serializer.descriptor)
```

File: runtime/commonMain/src/kotlinx/serialization/json/internal/TreeJsonOutput.kt [hunk 7]
```diff
@@ -68,8 +56,13 @@ private sealed class AbstractJsonTreeOutput(
     }
 
     override fun <T> encodeSerializableValue(serializer: SerializationStrategy<T>, value: T) {
-        encodePolymorphically(serializer, value) {
-            writePolymorphic = true
+        if (serializer.descriptor.kind !is PrimitiveKind && serializer.descriptor.kind !== UnionKind.ENUM_KIND || currentTagOrNull != null) {
```
Comment (qwwdfsad) [inline runtime/commonMain/src/kotlinx/serialization/json/internal/TreeJsonOutput.kt]: Please leave a short comment why this check is here (especially `currentTagOrNull`)

**Comment by qwwdfsad** [comment]
Also, it closes #287

File: runtime/commonMain/src/kotlinx/serialization/internal/Enums.kt (hunk 1 above)
Comment (sandwwraith) [reply runtime/commonMain/src/kotlinx/serialization/internal/Enums.kt]: It would be convenient if someone would be willing to create this serializer manually. However, I'm not sure that this use-case will be very popular

File: runtime/commonMain/src/kotlinx/serialization/json/internal/TreeJsonOutput.kt (hunk 4 above)
Comment (sandwwraith) [reply runtime/commonMain/src/kotlinx/serialization/json/internal/TreeJsonOutput.kt]: Well, then `JsonPrimitiveInput` should be refactored as well. I think I can share the constant between these classes

File: runtime/commonMain/src/kotlinx/serialization/json/internal/TreeJsonOutput.kt (hunk 6 above)
Comment (sandwwraith) [reply runtime/commonMain/src/kotlinx/serialization/json/internal/TreeJsonOutput.kt]: Well, actually, calls to `encode` change output state somehow, so I don't see problems with `apply`. Maybe something like this?

```kotlin
if (/**...**/) {
    encodePolymorphically(serializer, value) { writePolymorphic = true }
} else JsonTreePrimitiveOutput(json, nodeConsumer).apply {
    encodeSerializableValue(serializer, value)
    endEncode(serializer.descriptor)
}
```

File: runtime/commonMain/src/kotlinx/serialization/internal/Enums.kt (hunk 1 above)
Comment (qwwdfsad) [reply runtime/commonMain/src/kotlinx/serialization/internal/Enums.kt]: Then it's better not to introduce it in public API I suppose

**Comment by sandwwraith** [comment]
Merged manually
